src/Tokenizer/Analyzer/FunctionsAnalyzer.php

```diff
@@ -33,59 +34,52 @@ final class FunctionsAnalyzer
 
     /**
      * Important: risky because of the limited (file) scope of the tool.
+     *
+     * Faster version of isGlobalFunctionCall() which allows to pass in the namespace and therefore pre-compute it.
+     *
+     * @param list<NamespaceAnalysis> $namespaceDeclarations
      */
-    public function isGlobalFunctionCall(Tokens $tokens, int $index): bool
+    public function isGlobalFunctionCallWithNamespaces(Tokens $tokens, int $index, array $namespaceDeclarations): bool
```

Comment:
Do we need separate method for that? Can't we make `$namespaceDeclarations` optional and align method's internal logic to work differently whether if it's passed or not? I don't see value in widening public API of this class, especially with method name taken from arguments' list (in one of companies where I've been working was a guy who's been naming method like that and it was a disaster, like `getSomethingFromSomethingUsingSomethingAndTheOtherThing($something, $otherThing)` 😅).

Also, it seems little weird to me that `$namespaceDeclarations` is passed explicitly, since it could be done internally using `(new NamespacesAnalyzer())->getDeclarations($tokens)` because tokens are passed here 🤔. Approach with separate arguments allows namespace information to be out of sync (pass tokens and namespaces taken from different tokens set).

Reply:
great suggestions. the diff is now straigt forward.

Reply:
Yeah, now diff is more friendly, and I properly see that namespace analyser was already used in `isGlobalFunctionCall()` 😅. 

When I look at it now, I am wondering if it would be a good idea to introduce `Tokens::getNamespaceDeclarations()`, which should make analysis and keep the result for further usage 🤔. It would resolve several problems:
- performance (run analysis only once* per `Tokens`)
- `$namespaceDeclarations` possibly being out-of-sync with `$tokens`
- redundant `NamespacesAnalyzer` usage spread across several files (`list<NamespaceAnalysis>` wouldn't be passed as an additional argument here and there)

It could simplify this PR even more. What do you think @staabm? CC: @kubawerlos 

\* "once", because when `Tokens` instance is mutated (`insertAt()`, `clearAt()` and so on...) it should cause namespaces to be recalculated. This way `Tokens::getNamespaceDeclarations()` always would return up-to-date list of namespaces, but without doing the same thing multiple times.

Reply:
> $namespaceDeclarations possibly being out-of-sync with $tokens

how do we know when the cache in `Tokens::getNamespaceDeclarations()` needs to be cleared?

Reply:
> how do we know when the cache in Tokens::getNamespaceDeclarations() needs to be cleared?

That's the tricky part 😅. One possible option is to explicitly clear calculated namespaces on every mutating action, but it's bug-prone since it must be added to every potential, new mutating method.. Maybe there's some way to calculate some hash of current tokens set and check if namespace list was calculated for this exact set, and if not - re-calculate it 🤔. But that hash calculation (or any other way of determining changes) must be good from performance point of view, so whole change in analysing namespaces makes sense 😉.

Reply:
I just added a `$this->namespaceDeclarations = null;` after each `$this->changed = true`

Reply:
I believe it's good enough for now, it can be refactored and improved later so there's easier way of determining change in the tokens 👍.